content/posts/creating-grafana-dashboards-for-nodejs-on-kubernetes/index.md

```diff
@@ -0,0 +1,261 @@
+
+![Metric configuration for the heatmap](./heatmap-config-metric.png)
+
+> Important!
+>
+> In the panel section of the Grafana panel, on the `Axis` subsection, make sure to select `Time series buckets` in the `Data Format` field.
+
+Your heatmap should now be properly displayed.
+
+I recommend you hide 0 values and experiment with the colors. In the example, it was useful to chose `opacity` as the color mode and use `sqrt` as the scale to increase contrast of bucket counts with low values.
+
+### The Summary Metric Type
+
+The last example should have helped you get familiar with the histogram metric type.
+
+We now turn to the summary metric type which concludes the 4 basic metric types in Prometheus.
+
+A summary metric is similar to a histogram, as it is also a composite of multiple single metrics, but it is different, in that it calculates a set of quantiles (e.g. the 50th, 90th, 95th and 99th percentiles) exactly. This type of metric is more complex to calculate for export but is the best way to get exact quantiles, when you need them.
+
+Prometheus also allows you to calculate quantiles from histograms using the [histogram_quantile() function](https://prometheus.io/docs/prometheus/latest/querying/functions/#histogram_quantile), but this function uses interpolation to calculate the quantiles and therefore is not exact.
+
+## Monitoring Your Service Level Objectives
+
+A service level objective is often stated by requiring that a percentage of requests over a certain time period are shorter than a specific threshold, for exampe "99 percent of requests must be shorter than 0.3 seconds over any 24 hour period". Usually the time period is measured in days (as in the example), weeks, or months. In our example, the percentage of requests shorter than 0.3 seconds over 24 hours would be called a service level indicator (SLI), and we want to show it on our dashboard using the following query.
+
+```promql
+(sum by (app) (increase(http_request_duration_seconds_bucket{app="awesome-node", le="0.3"}[1d])) / sum by (app) (increase(http_request_duration_seconds_count{app="awesome-node"}[1d]))) * 100
+```
+
+In this case we are interested in a single number, therefore we want to use a stat panel.
+
+![Single stat panel showing the service level indicator for requests below 0.3 seconds](./service-level-objective.png)
+
+Below the number the stat can also show a graph giving you an indication of whether your SLI is heading up or down, or keeping steady.
+
+The single stat can be configured to change colors based on certain thresholds in the `Thresholds` section of the `Field` tab in Grafana. In this case I configured it to be red, when the SLI is below 99%, yellow, when it's lower than 99.9%, and green only if it's greater than or equal to 99.9%. You can see these thresholds are applied in the screenshot above, where the value 99.76% for our SLI is between 99% and 99.9% and therefore turned yellow.
```

Comment:
Since these are not pointers to code, I would just use double quotes instead of backticks to highlight them.